lib/dynamoid/transaction/executor.rb

```diff
@@ -0,0 +1,78 @@
+
+require 'dynamoid/transaction/action'
+require 'dynamoid/transaction/create'
+require 'dynamoid/transaction/delete'
+require 'dynamoid/transaction/update'
+require 'dynamoid/transaction/upsert'
+
+module Dynamoid
+  module Transaction
+    class Executor
+      def initialize(actions = [], options = {}, &block)
+        @skip_callbacks = !!options[:skip_callbacks]
+        @skip_validation = !!options[:skip_validation]
+        @actions = actions
+        instance_eval &block if block
+      end
+
+      def call
+        validate! unless @skip_validation
+        call_with_callbacks(@actions)
+      end
+
+      # throws Dynamoid::Errors::DocumentNotValid if model is present and not valid
+      # use skip_validation: true to skip
+      def validate!
+        @actions.each do |action|
+          next if !action.is_a?(Action) || !(model = action.model) || action.skip_validation?
+          raise Dynamoid::Errors::DocumentNotValid.new(model) unless model.valid?
+        end
+      end
+
+      # recursively call save, update, create, and validate callbacks
+      # only 'before' and first half of 'around' filters happen before transaction
+      def call_with_callbacks(actions = [])
+        return call_without_callbacks if actions.empty? # no models left so recursion ends
+        actions = actions.clone
+        action = actions.shift # remove first item
+        if action.respond_to?(:run_callbacks)
+          action.run_callbacks do
+            call_with_callbacks(actions) # recursive call with remaining items
+          end
+        else
+          call_with_callbacks(actions)
+        end
+      end
```

Comment:
Recursion is used so `before` callbacks are run on every model, then models are persisted, then all `after` callbacks are run.
```
Transaction do
  save(user)
  save(tag)
end
```
unwinds to:
before `user` callbacks
before `tag` callbacks
persist
after `tag` callbacks
after `user` callbacks

Reply:
Ah, I see. Right 👍

Reply:
Seems fine.

It makes sense also to have `after_commit` and `after_rollback` but it could be added later.

Reply:
> This is how ActiveRecord does it

The only difference - in case of Rails persisted changes are visible and can be read in any `after_` callback. But with the current approach in this PR any `after_` callback will not read the changes because they aren't still persisted. It may be confusing.